lib/pymedphys/_experimental/streamlit/utilities/icom.py

```diff
@@ -0,0 +1,301 @@
+    )
+    _adjust_icom_datetime_to_remove_duplicates(icom_datetime)
+
+    meterset = pd.Series(
+        [pmp_icom_extract.extract(item, "Delivery MU")[1] for item in icom_data_points],
+        name="meterset",
+    )
+
+    machine_id = pd.Series(
+        [pmp_icom_extract.extract(item, "Machine ID")[1] for item in icom_data_points],
+        name="machine_id",
+    )
+
+    return icom_datetime, meterset, machine_id
+
+
+def _adjust_icom_datetime_to_remove_duplicates(icom_datetime):
+    _, unique_index, unique_counts = np.unique(
+        icom_datetime, return_index=True, return_counts=True
+    )
+
+    for index, count in zip(unique_index, unique_counts):
+        if count > 1:
+            time_delta = datetime.timedelta(seconds=1 / count)
+            for current_duplicate, icom_index in enumerate(
+                range(index + 1, index + count)
+            ):
+                icom_datetime.iloc[icom_index] += time_delta * (current_duplicate + 1)
+
+
+# TODO: Remove "allow_output_mutation" once determine what is causing
+# the issue here.
+@st.cache(show_spinner=False, allow_output_mutation=True)
+def get_icom_dataset(filepath):
+    icom_stream = read_icom_log(filepath)
+    icom_data_points = pmp_icom_extract.get_data_points(icom_stream)
+
+    icom_datetime, meterset, machine_id = get_icom_datetimes_meterset_machine(filepath)
+
+    raw_delivery_items = pd.DataFrame(
+        [pmp_icom_delivery.get_delivery_data_items(item) for item in icom_data_points],
+        columns=["meterset", "gantry", "collimator", "mlc", "jaw"],
+    )
+    if np.all(meterset != raw_delivery_items["meterset"]):
+        raise ValueError("Expected meterset extractions to agree.")
+
+    turn_table = pd.Series(
+        [
+            pmp_icom_extract.extract(item, "Table Isocentric")[1]
+            for item in icom_data_points
+        ],
+        name="turn_table",
+    )
+
+    energy = pd.Series(
+        [pmp_icom_extract.extract(item, "Energy")[1] for item in icom_data_points],
+        name="energy",
+    )
+
+    interlocks = pd.Series(
+        [pmp_icom_extract.extract(item, "Interlocks")[1] for item in icom_data_points],
+        name="interlocks",
+    )
+
+    beam_timer = pd.Series(
+        [pmp_icom_extract.extract(item, "Beam Timer")[1] for item in icom_data_points],
+        name="beam_timer",
+    )
+
+    width, length, centre_x, centre_y = _determine_width_length_centre(
+        raw_delivery_items["mlc"], raw_delivery_items["jaw"]
+    )
+
+    icom_dataset = pd.concat(
+        [
+            icom_datetime,
+            machine_id,
+            energy,
+            width,
+            length,
+            raw_delivery_items[["meterset", "gantry", "collimator"]],
+            turn_table,
+            interlocks,
+            beam_timer,
+            centre_x,
+            centre_y,
+        ],
+        axis=1,
+    )
+
+    return icom_dataset
+
+
+def _check_for_consistent_mlc_width_return_mean(weighted_mlc_positions):
```

Comment:
rather than using the word "check", perhaps "validate"?
Check often implies one will get a true/false return value.